Reorganize downscaling workflow & generate design points for herbaceous crops (#1)
ccmmf/downscaling: docs/workflow_documentation.md

---
title: "Downscaling Workflow Documentation"
author: "David LeBauer"
date: "`r Sys.Date()`"
format:
  html:
    self-contained: true
    embed-resources: true
    toc: true
execute:
  echo: false
---

# Workflow Overview

The downscaling workflow predicts carbon pools (Soil Organic Carbon and Aboveground Biomass) for cropland fields in California and then aggregates these predictions to the county scale.

It uses an ensemble-based approach to uncertainty propagation and analysis, maintaining ensemble structure to propagate errors through the prediction and aggregation processes.

## Terminology

- **Design Points**: Fields chosen via stratified random sampling using k-means clustering on environmental data layers across California crop fields.
- **Crop Fields**: All croplands in the LandIQ dataset.
- **Anchor Sites:** Sites used as ground truth for calibration and validation, including UC research stations and Ameriflux sites.

## Workflow Steps

### Configuration

> Workflow settings are configured in `000-config.R`, except that the CCMMF_DIR is set in `.Renviron`. 

Reviewer: Why the exception for CCMMF_DIR? if the idea of the config script is to collect all the variables in one place, seems like the base dir might as well live there too.

Author: Right now I can't recall 
```suggestion
Workflow settings are configured in `000-config.R`, except that the CCMMF_DIR is set in `.Renviron`. 
<!-- TODO: check why CCMMF_DIR is handled separately; 
     I think there were two related motivations
        1. .Renviron vars can be overridden by `export CCMMF_DIR=...` 
        2. renv directories are set there, using the CCMMF_DIR variable. 
           but also not sure why those wouldn't be in there.
```

The configuration script reads the CCMMF directory from the environment variable `CCMMF_DIR` (set in .Renviron), and uses it to define paths for inputs and outputs.

#### Configuration setup

> **Required changes**

Reviewer: Can you add a few more words of explanation here? Are these a set of changes the user needs to make before every run, or to be done once at repo setup, or a TODO list for us  as the developers, or...?

Author: ```suggestion

To set up this workflow to run on your system, follow the following steps.

**Clone Repository**

```sh
git clone git@github.com:ccmmf/downscaling
```

**Set required system-specific parameters**

```

- `.Renviron` 
  - `CCMMF_DIR` should point to the shared CCMMF directory. 
    This is the location where data, inputs, and outputs will be stored, 
    as well as the location of the `renv` cache and library
  - `RENV_PATHS_CACHE` and `RENV_PATHS_LIBRARY` store the `renv` cache and library in the CCMMF directory.
    These are in a subdirectory of the CCMMF directory in order to make them available across all users 
    (and because on some computers, they exceed allocated space in the home directory).
- `000-config.R`
  - set `pecan_outdir` based on the CCMMF_DIR.
  - confirm that relative paths (`data_raw`, `data`, `cache`) are correct.
- For testing, keep `PRODUCTION` set to `FALSE`. This is _much_ faster and 

>   requires fewer computing resources. Once a test run is successful,

Author: ```suggestion
  requires fewer computing resources because it subsets large datasets. 
  Once a test run is successful,
```

>   set `PRODUCTION` to `TRUE` to run the full workflow.

Reviewer: Good to know it's faster, but what does this flag actually change?

Author: Added above

**Others:**

_these shouldn't need to be changed unless you want to change the default behavior of the workflow_

- `.future.R` defines parallel processing when the future package is loaded. 
  - set to `available cores - 1` by default
- `renv.lock` is used for package management with `renv`. 
See [project renv setup docs](renv_setup.md) for instructions about using `renv` for these workflows. 
See [renv package documentation](https://rstudio.github.io/renv/articles/renv.html) for more details.

**UdUnits dependency**

> 

Author: ```suggestion

If you get an error installing the units package, this - or something similar - may help. 
We are working on an alternative to renv that will bundle system dependencies and hopefully make this and related challenges unnecessary.
```

install units package

```r
install.packages(
  "units",
  configure.args = "--with-udunits2-lib=/share/pkg.8/udunits/2.2.28/install/lib --with-udunits2-include=/share/pkg.8/udunits/2.2.28/install/include"
)

> ```

Reviewer: This needs to be explained in more depth including some why along with the how: It's not clear to me from this what functionality you need it for and what problem these configure args are solving -- which means I wouldn't have a way of determining how to adapt these instructions to other systems.

Also I see geo already has units 0.8.5 in the R 4.4 system library -- would it have been possible to use that?

Author: When I was setting up renv it wanted to re-install all libraries, and I ran into this and related trouble. I have fiddled with renv so that it now does not install all packages if it is in the system library. @hdpriest-ui has a solution that may make renv and associated hoop jumping unnecessary. I'll add a clarifying comment.

### 1. Data Preparation

```sh
Rscript scripts/009_update_landiq.R
Rscript scripts/010_prepare_covariates.R
Rscript scripts/011_prepare_anchor_sites.R
```

> This script prepares data for clustering and downscaling:

Reviewer: not clear if "this script" refers to 009, 010, 011, or all three together

Author: ```suggestion
These scripts prepare data for clustering and downscaling:
```

- Converts LandIQ-derived shapefiles to a geopackage with geospatial information and a CSV with other attributes
- Extracts environmental covariates (clay, organic carbon, topographic wetness, temperature, precipitation, solar radiation, vapor pressure)
- Groups fields into Cal-Adapt climate regions
- Assigns anchor sites to fields

**Inputs:**

- **LandIQ Crop Map**: `data_raw/i15_Crop_Mapping_2016_SHP/i15_Crop_Mapping_2016.shp`

> - **Soilgrids**: `clay_0-5cm_mean.tif` and `ocd_0-5cm_mean.tif`

Reviewer: Using only the top 5-cm layer seems likely to be noisy -- have you compared results with  clay and ocd composited from  0-30 cm?

Reviewer: (Sorry if I asked this before; if so I don't remember the answer)

Author: I haven't - I am still focusing on MVP but will add a note and create an issue to explore this in the future.

```suggestion
- **Soilgrids**: `clay_0-5cm_mean.tif` and `ocd_0-5cm_mean.tif`
  Consider aggregating 0-5,5-15,15-30 into a single 0-30 cm layer 
```

| clay | Clay content | SoilGrids | % |
| ocd | Organic carbon density | SoilGrids | g/kg |
| twi | Topographic wetness index | SRTM-derived | - |

### 2. Design Point Selection

```sh
Rscript scripts/020_cluster_and_select_design_points.R
Rscript scripts/021_clustering_diagnostics.R
```

Uses k-means clustering to select representative fields plus anchor sites:

- Subsample LandIQ fields and include anchor sites for clustering
- Select cluster number based on the Elbow Method
- Cluster fields using k-means based on environmental covariates
- Select design points from clusters for SIPNET simulation

**Inputs:**
- `data/site_covariates.csv`
- `data/anchor_sites_ids.csv`

**Output:** 
- `data/design_points.csv`

### 3. SIPNET Model Runs

A separate workflow prepares inputs and runs SIPNET simulations for the design points.

**Inputs:**
- `design_points.csv`
- Initial conditions (from modeling workflow)

**Outputs:**
- `out/ENS-<ensemble_number>-<site_id>/YYYY.nc`: NetCDF files containing SIPNET outputs, in PEcAn standard model output format. 
- `out/ENS-<ensemble_number>-<site_id>/YYYY.nc.var`: List of variables included in SIPNET output (see table below)

 **Available Variables**

> Each output file named `YYYY.nc` contains an associated file named `YYYY.nc.var`. 

Reviewer: contains, or exists as a sibling?

Author: ```suggestion
Each output file named `YYYY.nc` has an associated file named `YYYY.nc.var`. 
```

 | SWE                           | Snow Water Equivalent                    |
 | litter_carbon_content         | Litter Carbon Content                    |
 | litter_mass_content_of_water  | Average layer litter moisture            |
 | LAI                           | Leaf Area Index                          |
 | fine_root_carbon_content      | Fine Root Carbon Content                 |
 | coarse_root_carbon_content    | Coarse Root Carbon Content               |
 | GWBI                          | Gross Woody Biomass Increment            |
 | AGB                           | Total aboveground biomass                |
 | time_bounds                   | history time interval endpoints          |

### 4. Extract SIPNET Output

```sh
Rscript scripts/030_extract_sipnet_output.R
```

Extracts and formats SIPNET outputs for downscaling:

- Extract output variables (AGB, TotSoilCarb) from SIPNET simulations
- Aggregate site-level ensemble outputs into long and 4D array formats
- Save CSV and NetCDF files following EFI standards

**Inputs:**
- `out/ENS-<ensemble_number>-<site_id>/YYYY.nc`

**Outputs:**
- `out/efi_ens_long.csv`: Long format data

> - `out/efi_forecast.nc`: NetCDF arrays

Reviewer: I remember asking for these to be renamed in the last round -- let's please name files for their contents, not for the format they use. 

Additionally, are these two distinct outputs or one result saved in two different formts? If the latter, are we actually using both of them?

Author: I did rename them. Will remove from this doc

author: "David LeBauer"
date: "`r Sys.Date()`"
format:
  html:
    self-contained: true
    embed-resources: true
    toc: true
execute:
  echo: false
---

# Workflow Overview

The downscaling workflow predicts carbon pools (Soil Organic Carbon and Aboveground Biomass) for cropland fields in California and then aggregates these predictions to the county scale.

It uses an ensemble-based approach to uncertainty propagation and analysis, maintaining ensemble structure to propagate errors through the prediction and aggregation processes.

![Spatial downscaling workflow using machine learning with environmental covariates](figures/spatial_downscaling_workflow.png){width="5in"}

## Terminology

- **Design Points**: Fields chosen via stratified random sampling using k-means clustering on environmental data layers across California crop fields.
- **Crop Fields**: All croplands in the LandIQ dataset.
- **Anchor Sites:** Sites used as ground truth for calibration and validation, including UC research stations and Ameriflux sites.

## Workflow Steps

### Configuration

Workflow settings are configured in `000-config.R`, except that the CCMMF_DIR is set in `.Renviron`. 

> <!-- TODO: check why CCMMF_DIR is handled separately; 

Reviewer: Not that you necessarily want a third option, but in the workflows repo I've converged on passing all configuration variables as command-line arguments with defaults set via {optparse}. It's verbose but highly discoverable, allows overrides for any one step if needed, and forced me to actually write the help strings for everything.
